src/algorithm/radix4.rs

```diff
@@ -129,34 +146,60 @@ impl<T: FftNum> Radix4<T> {
 }
 boilerplate_fft_oop!(Radix4, |this: &Radix4<_>| this.len);
 
-// after testing an iterative bit reversal algorithm, this recursive algorithm
-// was almost an order of magnitude faster at setting up
-fn prepare_radix4<T: FftNum>(
-    size: usize,
-    base_len: usize,
-    signal: &[Complex<T>],
-    spectrum: &mut [Complex<T>],
-    stride: usize,
+// Preparing for radix 4 is similar to a transpose, where the column index is bit reversed.
+// Use a lookup table to avoid repeating the slow bit reverse operations.
+// Unrolling the outer loop by a factor 4 helps speed things up.
+pub unsafe fn bitreversed_transpose<T: Copy>(
+    height: usize,
+    input: &[T],
+    output: &mut [T],
+    shuffle_map: &[usize],
 ) {
-    if size == base_len {
-        unsafe {
-            for i in 0..size {
-                *spectrum.get_unchecked_mut(i) = *signal.get_unchecked(i * stride);
-            }
-        }
-    } else {
-        for i in 0..4 {
-            prepare_radix4(
-                size / 4,
-                base_len,
-                &signal[i * stride..],
-                &mut spectrum[i * (size / 4)..],
-                stride * 4,
-            );
+    let width = shuffle_map.len();
+    for x in 0..width / 4 {
+        let x0 = 4 * x;
+        let x1 = 4 * x + 1;
+        let x2 = 4 * x + 2;
+        let x3 = 4 * x + 3;
+        let x_rev0 = shuffle_map.get_unchecked(x0);
```

Comment:
I only have one piece of feedback, which is that I think it'll be very hard in practice to verify that x_rev0, x_rev1 etc cannot be out of bounds. For most variables that we do unsafe lookups for, it's just an incrementing loop variable times a stride, but this is the output of an algorithm that's pretty opaque to my intuition.

The same was true of the recursive algorithm I guess, but I think my own brain just visualizes recursion better. How much of a performance hit is it to make the `*output.get_unchecked_mut(output_index0) = temp0;` lines below to be checked? Even if there's a small performance loss, this might be a rare case where it's worth taking the hit.

Reply:
As a bonus, if those lookups are checked, then this function doesn't need to be unsafe, since all of its unsafety is internal.

Reply:
Also: Now that I'm looking into it more, I wonder: Do these lookups into `shuffle_map` have the performance if we use `chunks_exact(4)` instead of unsafe lookups? If so, the less unsafe the better.

Reply:
I'll try these things and see what happens!

Reply:
Checked indexing turned out to hurt pretty bad, it was up to 25% slower. But `chunks_exact(4)`on `shuffle_map` is fine, I switched to that. 
I also added some asserts that should give the same safety as the checked indexing. The asserts are in the outer loop, and there they only give a very minor (and acceptable I think) slowdown. Only a few lengths lose speed, and then only up to a few %.

Reply:
One last idea is to move those asserts into the constructor.

Reply:
Now that I type it out though, I think I like putting it in the outer loop though, because there's a lot of value in making the function itself safe